File: GuapoTerminal/app/src/main/java/com/guaposniper/terminal/MainActivity.java
package com.guaposniper.terminal;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.method.ScrollingMovementMethod;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {

    private TextView terminalOutput;
    private EditText terminalInput;
    private ScrollView scrollView;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private File currentDir;
    private StringBuilder outputBuffer = new StringBuilder();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        terminalOutput = findViewById(R.id.terminal_output);
        terminalInput = findViewById(R.id.terminal_input);
        scrollView = findViewById(R.id.scroll_view);

        terminalOutput.setMovementMethod(new ScrollingMovementMethod());
        currentDir = getFilesDir();

        appendOutput("╔══════════════════════════════════════╗\n");
        appendOutput("║       GuapoTerminal v1.0             ║\n");
        appendOutput("║   by GuapoSniper                     ║\n");
        appendOutput("║   Ton terminal perso Android          ║\n");
        appendOutput("╚══════════════════════════════════════╝\n\n");
        showPrompt();

        terminalInput.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE ||
                (event != null && event.getKeyCode() == KeyEvent.KEYCODE_ENTER
                 && event.getAction() == KeyEvent.ACTION_DOWN)) {
                String command = terminalInput.getText().toString().trim();
                if (!command.isEmpty()) {
                    executeCommand(command);
                    terminalInput.setText("");
                }
                return true;
            }
            return false;
        });
    }

    private void showPrompt() {
        String dirName = currentDir.getName();
        if (dirName.isEmpty()) dirName = "/";
        appendOutput("guapo@terminal:" + dirName + "$ ");
    }

    private void executeCommand(String command) {
        appendOutput(command + "\n");

        // Handle built-in commands
        if (command.equals("clear") || command.equals("cls")) {
            outputBuffer = new StringBuilder();
            mainHandler.post(() -> terminalOutput.setText(""));
            showPrompt();
            return;
        }

        if (command.equals("help")) {
            showHelp();
            showPrompt();
            return;
        }

        if (command.equals("exit")) {
            finish();
            return;
        }

        if (command.equals("pwd")) {
            appendOutput(currentDir.getAbsolutePath() + "\n");
            showPrompt();
            return;
        }

        if (command.startsWith("cd ")) {
            changeDirectory(command.substring(3).trim());
            showPrompt();
            return;
        }

        if (command.equals("whoami")) {
            appendOutput("guaposniper\n");
            showPrompt();
            return;
        }

        if (command.equals("neofetch")) {
            showNeofetch();
            showPrompt();
            return;
        }

        // Execute shell command
        executor.execute(() -> {
            try {
                ProcessBuilder pb = new ProcessBuilder("sh", "-c", command);
                pb.directory(currentDir);
                pb.redirectErrorStream(true);
                Process process = pb.start();

                BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream())
                );

                String line;
                while ((line = reader.readLine()) != null) {
                    final String output = line;
                    mainHandler.post(() -> appendOutput(output + "\n"));
                }

                process.waitFor();
                reader.close();

            } catch (Exception e) {
                mainHandler.post(() ->
                    appendOutput("Erreur: " + e.getMessage() + "\n")
                );
            }

            mainHandler.post(this::showPrompt);
        });
    }

    private void changeDirectory(String path) {
        File newDir;
        if (path.equals("..")) {
            newDir = currentDir.getParentFile();
            if (newDir == null) newDir = currentDir;
        } else if (path.startsWith("/")) {
            newDir = new File(path);
        } else {
            newDir = new File(currentDir, path);
        }

        if (newDir.exists() && newDir.isDirectory()) {
            currentDir = newDir;
        } else {
            appendOutput("cd: " + path + ": Dossier introuvable\n");
        }
    }

    private void showHelp() {
        appendOutput("\n");
        appendOutput("╔══ GuapoTerminal - Commandes ══╗\n");
        appendOutput("║                                ║\n");
        appendOutput("║  help     - Afficher l'aide    ║\n");
        appendOutput("║  clear    - Effacer l'ecran    ║\n");
        appendOutput("║  pwd      - Dossier actuel     ║\n");
        appendOutput("║  cd <dir> - Changer dossier    ║\n");
        appendOutput("║  ls       - Lister fichiers    ║\n");
        appendOutput("║  whoami   - Ton nom            ║\n");
        appendOutput("║  neofetch - Info systeme       ║\n");
        appendOutput("║  exit     - Quitter            ║\n");
        appendOutput("║                                ║\n");
        appendOutput("║  + toutes les commandes shell  ║\n");
        appendOutput("╚════════════════════════════════╝\n\n");
    }

    private void showNeofetch() {
        String model = android.os.Build.MODEL;
        String brand = android.os.Build.BRAND;
        String version = android.os.Build.VERSION.RELEASE;
        int sdk = android.os.Build.VERSION.SDK_INT;

        appendOutput("\n");
        appendOutput("   ██████╗ ██╗   ██╗ █████╗ ██████╗  ██████╗ \n");
        appendOutput("  ██╔════╝ ██║   ██║██╔══██╗██╔══██╗██╔═══██╗\n");
        appendOutput("  ██║  ███╗██║   ██║███████║██████╔╝██║   ██║\n");
        appendOutput("  ██║   ██║██║   ██║██╔══██║██╔═══╝ ██║   ██║\n");
        appendOutput("  ╚██████╔╝╚██████╔╝██║  ██║██║     ╚██████╔╝\n");
        appendOutput("   ╚═════╝  ╚═════╝ ╚═╝  ╚═╝╚═╝      ╚═════╝\n");
        appendOutput("  ─────────────────────────────────────\n");
        appendOutput("  OS:      Android " + version + " (SDK " + sdk + ")\n");
        appendOutput("  Device:  " + brand + " " + model + "\n");
        appendOutput("  Shell:   GuapoTerminal v1.0\n");
        appendOutput("  User:    guaposniper\n");
        appendOutput("  ─────────────────────────────────────\n\n");
    }

    private void appendOutput(String text) {
        outputBuffer.append(text);
        terminalOutput.setText(outputBuffer.toString());
        scrollView.post(() -> scrollView.fullScroll(ScrollView.FOCUS_DOWN));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        executor.shutdown();
    }
}

File: app/src/main/java/com/guaposniper/terminal/MainActivity.kt
package com.guaposniper.terminal

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.ImageButton
import android.widget.ScrollView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

class MainActivity : AppCompatActivity() {

    private lateinit var terminalOutput: TextView
    private lateinit var commandInput: EditText
    private lateinit var sendButton: ImageButton
    private lateinit var scrollView: ScrollView
    private lateinit var promptLabel: TextView
    private lateinit var session: TerminalSession

    private val outputBuffer = SpannableStringBuilder()

    companion object {
        private const val STORAGE_PERMISSION_CODE = 100
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        terminalOutput = findViewById(R.id.terminalOutput)
        commandInput = findViewById(R.id.commandInput)
        sendButton = findViewById(R.id.sendButton)
        scrollView = findViewById(R.id.scrollView)
        promptLabel = findViewById(R.id.promptLabel)

        session = TerminalSession()

        requestStoragePermission()
        showWelcomeBanner()
        updatePrompt()

        sendButton.setOnClickListener { executeCurrentCommand() }

        commandInput.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                executeCurrentCommand()
                true
            } else {
                false
            }
        }

        commandInput.setOnKeyListener { _, keyCode, event ->
            if (event.action == KeyEvent.ACTION_DOWN) {
                when (keyCode) {
                    KeyEvent.KEYCODE_DPAD_UP -> {
                        session.getPreviousCommand()?.let { commandInput.setText(it) }
                        commandInput.setSelection(commandInput.text.length)
                        true
                    }
                    KeyEvent.KEYCODE_DPAD_DOWN -> {
                        session.getNextCommand()?.let { commandInput.setText(it) }
                        commandInput.setSelection(commandInput.text.length)
                        true
                    }
                    else -> false
                }
            } else {
                false
            }
        }

        commandInput.requestFocus()
    }

    private fun showWelcomeBanner() {
        val banner = buildString {
            appendLine()
            appendLine("  ██████╗ ██╗   ██╗ █████╗ ██████╗  ██████╗ ")
            appendLine("  ██╔════╝ ██║   ██║██╔══██╗██╔══██╗██╔═══██╗")
            appendLine("  ██║  ███╗██║   ██║███████║██████╔╝██║   ██║")
            appendLine("  ██║   ██║██║   ██║██╔══██║██╔═══╝ ██║   ██║")
            appendLine("  ╚██████╔╝╚██████╔╝██║  ██║██║     ╚██████╔╝")
            appendLine("   ╚═════╝  ╚═════╝ ╚═╝  ╚═╝╚═╝      ╚═════╝")
            appendLine("  ████████╗███████╗██████╗ ███╗   ███╗")
            appendLine("  ╚══██╔══╝██╔════╝██╔══██╗████╗ ████║")
            appendLine("     ██║   █████╗  ██████╔╝██╔████╔██║")
            appendLine("     ██║   ██╔══╝  ██╔══██╗██║╚██╔╝██║")
            appendLine("     ██║   ███████╗██║  ██║██║ ╚═╝ ██║")
            appendLine("     ╚═╝   ╚══════╝╚═╝  ╚═╝╚═╝     ╚═╝")
            appendLine()
            appendLine("  ★ GuapoTerminal v1.0 - by GuapoSniper ★")
            appendLine("  Tape 'help' pour voir les commandes.")
            appendLine()
        }
        appendToOutput(banner, Color.parseColor("#00FFCC"))
    }

    private fun executeCurrentCommand() {
        val command = commandInput.text.toString()
        commandInput.setText("")

        val prompt = session.getPrompt()
        appendToOutput("$prompt $command\n", Color.parseColor("#00FFCC"))

        val result = session.execute(command)

        when {
            result.clearScreen -> {
                outputBuffer.clear()
                terminalOutput.text = ""
                showWelcomeBanner()
            }
            result.exitApp -> {
                finish()
            }
            result.output.isNotEmpty() -> {
                val color = if (result.isError) {
                    Color.parseColor("#FF4444")
                } else {
                    Color.parseColor("#E0E0E0")
                }
                appendToOutput(result.output + "\n", color)
            }
        }

        updatePrompt()
        scrollToBottom()
    }

    private fun appendToOutput(text: String, color: Int) {
        val start = outputBuffer.length
        outputBuffer.append(text)
        outputBuffer.setSpan(
            ForegroundColorSpan(color),
            start,
            outputBuffer.length,
            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )
        terminalOutput.text = outputBuffer
    }

    private fun updatePrompt() {
        promptLabel.text = session.getPrompt()
    }

    private fun scrollToBottom() {
        scrollView.post {
            scrollView.fullScroll(ScrollView.FOCUS_DOWN)
            commandInput.requestFocus()
        }
    }

    private fun requestStoragePermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            if (!android.os.Environment.isExternalStorageManager()) {
                try {
                    val intent = android.content.Intent(
                        android.provider.Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION
                    )
                    startActivity(intent)
                } catch (_: Exception) { }
            }
        } else {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE
                    ),
                    STORAGE_PERMISSION_CODE
                )
            }
        }
    }
}
